src/streamlink/plugins/vimeo.py

```diff
@@ -8,70 +8,175 @@
+                    validate.optional("hls"): {"cdns": {validate.text: {"url": validate.any(None, validate.url())}}},
                     validate.optional("progressive"): validate.all(
-                        [{"url": validate.url(), "quality": validate.text}]
+                        [{"url": validate.url(), "quality": validate.text}],
                     ),
                 },
                 validate.optional("text_tracks"): validate.all(
-                    [{"url": validate.text, "lang": validate.text}]
+                    [{"url": validate.text, "lang": validate.text}],
                 ),
-            }
+            },
         },
     )
-    _player_schema = validate.Schema(
-        validate.transform(_config_re.search),
-        validate.any(None, validate.Schema(validate.get(1), _config_schema)),
-    )
 
+    def __init__(self, session, url):
+        self.session = session
+        self.url = url
+        self._auth_data = None
+        self._last_updated = None
+        self._updating = False
+
+    def get_player_data(self):
+        return self.session.http.get(self.url, schema=validate.Schema(
+            validate.transform(re.compile(r"var\s+config\s*=\s*({.+?})\s*;").search),
+            validate.any(None, validate.Schema(validate.get(1), VimeoAPI._config_schema)),
+        ))
+
+    def get_api_url(self):
+        return self.session.http.get(self.url, schema=validate.Schema(
+            validate.transform(re.compile(r'(?:"config_url"|\bdata-config-url)\s*[:=]\s*(".+?")').search),
+            validate.any(
+                None,
+                validate.Schema(
+                    validate.get(1),
+                    validate.parse_json(),
+                    validate.transform(html_unescape),
+                    validate.url(),
+                ),
+            ),
+        ))
+
+    def get_config_data(self, api_url):
+        return self.session.http.get(api_url, schema=VimeoAPI._config_schema)
+
+
+@pluginmatcher(re.compile(
+    r"https?://(player\.vimeo\.com/video/\d+|(www\.)?vimeo\.com/.+)"
+))
+class Vimeo(Plugin):
     arguments = PluginArguments(
         PluginArgument("mux-subtitles", is_global=True)
     )
 
-    def _get_streams(self):
-        if "player.vimeo.com" in self.url:
-            data = self.session.http.get(self.url, schema=self._player_schema)
+    def _get_data(self):
+        if "player.vimeo.com" in self.session.api.url:
+            return self.session.api.get_player_data()
         else:
-            api_url = self.session.http.get(self.url, schema=self._config_url_schema)
+            api_url = self.session.api.get_api_url()
             if not api_url:
                 return
-            data = self.session.http.get(api_url, schema=self._config_schema)
+            return self.session.api.get_config_data(api_url)
+
+    def _get_streams(self):
+        self.session.api = VimeoAPI(self.session, self.url)
```

Comment:
Don't add new attributes to the session object. You can pass an instance of the VimeoAPI class to the respective stream constructors.

Reply:
The reason I did this was to get around a problem.  If I try to pass `api` as a keyword to `parse_variant_playlist()`, I end up with:
```
$ streamlink https://vimeo.com/720864225 240p
[cli][info] Found matching plugin vimeo for URL https://vimeo.com/720864225
error: Unable to open URL: https://live-api.vimeocdn.com/sessions/c5c51d94-5805-44c1-96ef-72f050453867/playlist/hls.m3u8?cdn=akamai-low-latency&~id=player&~exp=1655420400&~sig=kDTUIsLrnH1jAKp8DsStXB30txhJu2LB7J-Ki6dltR8 (request() got an unexpected keyword argument 'api')
```

Reply:
If you add the `api` arg or keyword to the constuctor's signature, then it won't be included in the args keyword dict, and only relevant args will then be passed to requests afterwards.

```py
    def __init__(self, session_, url, api: "VimeoAPI", **args):
```

I've been working on refactoring the plugin, based on your PR, which cleans up all the issues and simplifies the URL refresh logic, but it's not yet finished and I still haven't taken a closer look at Vimeo itself, as all the videos I could find were "DASH" streams in some JSON format.

As said, the main problem with your current implementation is the API stuff and how you're refreshing the URLs. Since you're refreshing data once in one thread, but are accessing the url property in multiple threads, you will have to use thread locks to prevent timing issues, and also for waiting on the availability of the new URLs. The URL-data acquirement and evaluation should also be moved to the API class, and the HLSStream subclass should be kept as simple as possible.

Reply:
Hmm, it doesn't seem like it's worthwhile for me to continue this if you're refactoring the plugin in parallel.

Reply:
No, I was just curious and looking for a solution myself. This is nowhere near functional.

Reply:
> If you add the api arg or keyword to the constuctor's signature, then it won't be included in the args keyword dict, and only relevant args will then be passed to requests afterwards.

You're talking about the constructor for `VimeoHLSStream`, right?

Maybe I'm being a bit thick here, but I don't see how that solves the problem.  The problem is that the keyword args to `parse_variant_playlist()` are used directly for requests within the method.

The following is one possible solution that works for me:
```
--- hls.py.orig	2022-06-17 01:30:03.279042490 +0100
+++ hls.py	2022-06-17 01:23:21.410866145 +0100
@@ -632,10 +632,11 @@
                                or :py:meth:`requests.Session.request`
         """
 
+        reqargs = session_.http.valid_request_args(**request_params)
         locale = session_.localization
         audio_select = session_.options.get("hls-audio-select") or []
 
-        res = cls._fetch_variant_playlist(session_, url, **request_params)
+        res = cls._fetch_variant_playlist(session_, url, **reqargs)
 
         try:
             multivariant = cls._get_variant_playlist(res)
@@ -730,7 +731,7 @@
             if check_streams:
                 # noinspection PyBroadException
                 try:
-                    session_.http.get(playlist.uri, **request_params)
+                    session_.http.get(playlist.uri, **reqargs)
                 except KeyboardInterrupt:
                     raise
                 except Exception:
```

Reply:
> The problem is that the keyword args to parse_variant_playlist() are used directly for requests within the method.

You're right, I didn't see this, sorry.

Well, in order to fix this, and in order to do it properly, the signature of the `HLSStream.parse_variant_playlist` class method would need to get changed, which would be a breaking change. Not that it's something we need to avoid, but it's something that needs to be evaluated carefully. As I mentioned in other threads earlier, I wanted to do that anyway at some point.

The main issue is that any parameters meant for `requests` should not be passed as additional unknown keywords and should instead be defined as a specific dict keyword that's empty by default. Then any additional unknown keywords can be passed to the HLSStream/MuxedHLSStream or its subclasses, so stuff like `force_restart` doesn't have to be defined here.

The DASHStream.parse_manifest class method has the same problem, but it's not extensible at all and everything's hardcoded.

Reply:
OK, I'm just going to use the patch I posted above for the time being, so that I can pass the `api` keyword to `parse_variant_playlist()` to progress the vimeo plugin work.  I'll tidy up the other issues you have identified here (mostly done now) with the plugin.